File: 学校事务/backup/compiler-backend/src/main/java/com/huangbin/compiler/lexer/Lexer.java
package com.huangbin.compiler.lexer;

import com.huangbin.compiler.model.Diagnostic;
import com.huangbin.compiler.model.Token;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Lexer {
    private final String sourceCode;
    private int pos = 0;
    private int line = 1;
    private int column = 1;

    private final List<Token> tokens = new ArrayList<>();
    private final List<Diagnostic> diagnostics = new ArrayList<>();

    // 预定义词法字典
    private static final Map<String, Integer> KEYWORDS = new HashMap<>();
    private static final Map<String, Integer> OPERATORS = new HashMap<>();

    static {
        // 初始化关键字
        KEYWORDS.put("int", 101); KEYWORDS.put("float", 102); KEYWORDS.put("char", 103);
        KEYWORDS.put("void", 104); KEYWORDS.put("main", 105); KEYWORDS.put("if", 106);
        KEYWORDS.put("else", 107); KEYWORDS.put("while", 108); KEYWORDS.put("for", 109);
        KEYWORDS.put("read", 110); KEYWORDS.put("write", 111); KEYWORDS.put("const", 112);
        KEYWORDS.put("return", 113); KEYWORDS.put("break", 114); KEYWORDS.put("continue", 115);

        // 初始化双字符和单字符操作符
        OPERATORS.put("==", 201); OPERATORS.put("!=", 202); OPERATORS.put("<=", 203);
        OPERATORS.put(">=", 204); OPERATORS.put("&&", 205); OPERATORS.put("||", 206);
        OPERATORS.put("=", 207);  OPERATORS.put(">", 208);  OPERATORS.put("<", 209);
        OPERATORS.put("+", 210);  OPERATORS.put("-", 211);  OPERATORS.put("*", 212);
        OPERATORS.put("/", 213);  OPERATORS.put("%", 214);  OPERATORS.put("!", 215);

        // 初始化界符
        OPERATORS.put("(", 301); OPERATORS.put(")", 302); OPERATORS.put("{", 303);
        OPERATORS.put("}", 304); OPERATORS.put("[", 305); OPERATORS.put("]", 306);
        OPERATORS.put(";", 307); OPERATORS.put(",", 308);
    }

    public Lexer(String sourceCode) {
        this.sourceCode = sourceCode == null ? "" : sourceCode;
    }

    public List<Token> getTokens() { return tokens; }
    public List<Diagnostic> getDiagnostics() { return diagnostics; }

    // 辅助获取字符
    private char current() { return pos < sourceCode.length() ? sourceCode.charAt(pos) : '\0'; }
    private char peek() { return pos + 1 < sourceCode.length() ? sourceCode.charAt(pos + 1) : '\0'; }

    // 游标前进并维护行号列号
    private void advance() {
        if (current() == '\n') { line++; column = 1; }
        else { column++; }
        pos++;
    }

    // 【兼容性设置】：创建 Token 的工厂方法
    private void createToken(int startLine, int startCol, String text, String kind, int code) {
        /*
         * 注意：如果你的 Token 类是通过 new Token(line, col, text, kind, code) 实例化的，
         * 请将下面的代码替换为：tokens.add(new Token(startLine, startCol, text, kind, code));
         */
        Token t = new Token();
        t.setLine(startLine);
        t.setColumn(startCol);
        t.setText(text);
        t.setKind(kind);
        t.setCode(code);
        tokens.add(t);
    }

    public void tokenize() {
        while (pos < sourceCode.length()) {
            char ch = current();

            // 1. 跳过空白字符
            if (Character.isWhitespace(ch)) { advance(); continue; }

            // 2. 跳过单行注释 //
            if (ch == '/' && peek() == '/') {
                while (current() != '\n' && current() != '\0') advance();
                continue;
            }

            // 3. 跳过多行注释 /* */
            if (ch == '/' && peek() == '*') {
                advance(); advance();
                while (current() != '\0' && !(current() == '*' && peek() == '/')) advance();
                if (current() != '\0') { advance(); advance(); }
                continue;
            }

            // 4. 【核心修复】：精准拦截单引号和双引号包裹的字符串！
            if (ch == '"' || ch == '\'') {
                char quote = ch;
                int startCol = column;
                StringBuilder str = new StringBuilder();
                str.append(ch); // 存入开头引号
                advance();

                while (current() != '\0' && current() != quote) {
                    str.append(current());
                    advance();
                }

                if (current() == quote) {
                    str.append(current()); // 存入结尾引号
                    advance();
                } else {
                    diagnostics.add(new Diagnostic("lexer", line, "L001", "字符串缺少闭合引号"));
                }

                createToken(line, startCol, str.toString(), "string_literal", 800);
                continue;
            }

            // 5. 识别关键字和标识符
            if (Character.isLetter(ch) || ch == '_') {
                int startCol = column;
                StringBuilder sb = new StringBuilder();
                while (Character.isLetterOrDigit(current()) || current() == '_') {
                    sb.append(current());
                    advance();
                }
                String text = sb.toString();
                if (KEYWORDS.containsKey(text)) {
                    createToken(line, startCol, text, "keyword", KEYWORDS.get(text));
                } else {
                    createToken(line, startCol, text, "identifier", 400);
                }
                continue;
            }

            // 6. 识别数字 (支持小数)
            if (Character.isDigit(ch)) {
                int startCol = column;
                StringBuilder sb = new StringBuilder();
                boolean hasDot = false;
                while (Character.isDigit(current()) || current() == '.') {
                    if (current() == '.') {
                        if (hasDot) break; // 防止 1.2.3 这种非法数字
                        hasDot = true;
                    }
                    sb.append(current());
                    advance();
                }
                createToken(line, startCol, sb.toString(), hasDot ? "float_literal" : "int_literal", hasDot ? 501 : 500);
                continue;
            }

            // 7. 识别双字符操作符 (如 ==, <=, &&)
            String twoCharOp = "" + ch + peek();
            if (OPERATORS.containsKey(twoCharOp)) {
                createToken(line, column, twoCharOp, "operator", OPERATORS.get(twoCharOp));
                advance(); advance();
                continue;
            }

            // 8. 识别单字符操作符和界符 (如 +, =, ;)
            String oneCharOp = "" + ch;
            if (OPERATORS.containsKey(oneCharOp)) {
                createToken(line, column, oneCharOp, "operator", OPERATORS.get(oneCharOp));
                advance();
                continue;
            }

            // 9. 容错处理：未知字符报错并跳过
            diagnostics.add(new Diagnostic("lexer", line, "L003", "unknown character '" + ch + "' at column " + column));
            advance();
        }
    }
}

File: 学校事务/编译原理/12303070250黄彬/compiler/lexer.py
from typing import List, Tuple

from .models import Diagnostic, Token


KEYWORDS = {
    "const": 100,
    "int": 101,
    "float": 102,
    "char": 103,
    "void": 104,
    "return": 105,
    "if": 106,
    "else": 107,
    "while": 108,
    "do": 109,
    "for": 110,
    "break": 111,
    "continue": 112,
}

OPERATORS = {
    "==": 201,
    "!=": 202,
    "<=": 203,
    ">=": 204,
    "&&": 205,
    "||": 206,
    "=": 207,
    ">": 208,
    "<": 209,
    "+": 210,
    "-": 211,
    "*": 212,
    "/": 213,
    "!": 214,
    "%": 215,
}

SEPARATORS = {
    ";": 301,
    ",": 302,
    "(": 303,
    ")": 304,
    "{": 305,
    "}": 306,
    "[": 307,
    "]": 308,
}

IDENTIFIER_CODE = 700
INT_LITERAL_CODE = 401
FLOAT_LITERAL_CODE = 402
CHAR_LITERAL_CODE = 403
STRING_LITERAL_CODE = 404
SUPPORTED_CHAR_ESCAPES = {"n", "t", "r", "0", "'", "\\"}
SUPPORTED_STRING_ESCAPES = {"n", "t", "r", "0", '"', "\\"}


class Lexer:
    def tokenize(self, source: str) -> Tuple[List[Token], List[Diagnostic]]:
        tokens: List[Token] = []
        diagnostics: List[Diagnostic] = []
        i = 0
        line = 1
        column = 1

        while i < len(source):
            ch = source[i]

            if ch in " \t\r":
                i += 1
                column += 1
                continue

            if ch == "\n":
                i += 1
                line += 1
                column = 1
                continue

            if source.startswith("//", i):
                while i < len(source) and source[i] != "\n":
                    i += 1
                    column += 1
                continue

            if source.startswith("/*", i):
                start_line = line
                start_column = column
                i += 2
                column += 2
                closed = False
                while i < len(source):
                    if source.startswith("*/", i):
                        i += 2
                        column += 2
                        closed = True
                        break
                    if source[i] == "\n":
                        i += 1
                        line += 1
                        column = 1
                    else:
                        i += 1
                        column += 1
                if not closed:
                    diagnostics.append(Diagnostic("lexer", start_line, "L001", f"unclosed comment at column {start_column}"))
                continue

            if ch.isalpha() or ch == "_":
                start = i
                start_column = column
                while i < len(source) and (source[i].isalnum() or source[i] == "_"):
                    i += 1
                    column += 1
                text = source[start:i]
                if text in KEYWORDS:
                    tokens.append(Token(text, KEYWORDS[text], line, start_column, "keyword"))
                else:
                    tokens.append(Token(text, IDENTIFIER_CODE, line, start_column, "identifier"))
                continue

            if ch.isdigit():
                start = i
                start_column = column
                has_dot = False
                while i < len(source) and (source[i].isdigit() or source[i] == "."):
                    if source[i] == ".":
                        if has_dot:
                            break
                        has_dot = True
                    i += 1
                    column += 1
                text = source[start:i]
                code = FLOAT_LITERAL_CODE if has_dot else INT_LITERAL_CODE
                kind = "float_literal" if has_dot else "int_literal"
                tokens.append(Token(text, code, line, start_column, kind))
                continue

            if ch == "'":
                start = i
                start_column = column
                i += 1
                column += 1
                while i < len(source):
                    if source[i] == "\n" or source[i] == "'":
                        break
                    if source[i] == "\\":
                        i += 1
                        column += 1
                        if i >= len(source) or source[i] == "\n":
                            break
                    i += 1
                    column += 1
                if i < len(source) and source[i] == "'":
                    i += 1
                    column += 1
                    text = source[start:i]
                    payload = text[1:-1]
                    is_ordinary_char = len(payload) == 1 and payload != "\\"
                    is_escape_char = len(payload) == 2 and payload[0] == "\\" and payload[1] in SUPPORTED_CHAR_ESCAPES
                    if is_ordinary_char or is_escape_char or payload:
                        tokens.append(Token(text, CHAR_LITERAL_CODE, line, start_column, "char_literal"))
                    else:
                        diagnostics.append(Diagnostic("lexer", line, "L004", f"malformed char literal at column {start_column}"))
                else:
                    diagnostics.append(Diagnostic("lexer", line, "L002", f"unclosed char literal at column {start_column}"))
                continue

            if ch == '"':
                start = i
                start_column = column
                i += 1
                column += 1
                closed = False
                valid = True
                while i < len(source):
                    if source[i] == "\n":
                        break
                    if source[i] == '"':
                        closed = True
                        i += 1
                        column += 1
                        break
                    if source[i] == "\\":
                        i += 1
                        column += 1
                        if i >= len(source) or source[i] not in SUPPORTED_STRING_ESCAPES:
                            valid = False
                            break
                    i += 1
                    column += 1
                if closed and valid:
                    tokens.append(Token(source[start:i], STRING_LITERAL_CODE, line, start_column, "string_literal"))
                else:
                    diagnostics.append(Diagnostic("lexer", line, "L005", f"malformed string literal at column {start_column}"))
                continue

            two = source[i : i + 2]
            if two in OPERATORS:
                tokens.append(Token(two, OPERATORS[two], line, column, "operator"))
                i += 2
                column += 2
                continue

            if ch in OPERATORS:
                tokens.append(Token(ch, OPERATORS[ch], line, column, "operator"))
                i += 1
                column += 1
                continue

            if ch in SEPARATORS:
                tokens.append(Token(ch, SEPARATORS[ch], line, column, "separator"))
                i += 1
                column += 1
                continue

            diagnostics.append(Diagnostic("lexer", line, "L003", f"unknown character {ch!r} at column {column}"))
            i += 1
            column += 1

        return tokens, diagnostics
